### PR TITLE
Add highlight context to AST/parser exceptions

### PR DESCRIPTION
**Pull request by invisig0th**

### CODE REVIEW

File: synapse/lib/parser.py [hunk 1]
```diff
@@ -175,72 +194,64 @@ def subquery(self, meta, kids):
...
-        return node
+        kids = [self._parseJsonToken(k) for k in kids]
+        astinfo = self.metaToAstInfo(meta)
+        return s_ast.ExprDict(astinfo, kids=kids)
 
     @lark.v_args(meta=True)
     def trycatch(self, meta, kids):
         kids = self._convert_children(kids)
-        return s_ast.TryCatch(kids=kids)
+        astinfo = self.metaToAstInfo(meta)
```
Comment (Cisphyx) [inline synapse/lib/parser.py:228]: There's a bunch of these rules that can just be in the 1 to 1 conversion map, tossed up a PR cleaning them up here https://github.com/vertexproject/synapse/pull/3092

**Review by infime — APPROVED** [review]

**Comment by infime** [comment]
looking good from the optic side :+1:

**Review by Cisphyx — APPROVED** [review]

File: synapse/exc.py [hunk 2]
```diff
@@ -217,6 +217,26 @@ class ModAlreadyLoaded(SynErr): pass
...
 class NotMsgpackSafe(SynErr): pass
 
+class NoSuchForm(SynErr):
+
+    # new convention where exceptions with an init() class
+    # method may be called to enforce normalization of message
+    # contents based on variables that are relevant to the
+    # exception type...
+    @classmethod
+    def init(cls, name, mesg=None):
```
Comment (vEpiphyte) [inline synapse/exc.py:227]: We will need to ensure this design pattern survives pickling in the event that the exceptions are passed across multiprocessing pool.

File: synapse/lib/parser.py [hunk 3]
```diff
@@ -141,30 +144,46 @@ def __init__(self, text):
...
         # Keep the original text for error printing and weird subquery argv parsing
         self.text = text
-
-    @classmethod
-    def _convert_children(cls, children):
-        return [cls._convert_child(k) for k in children]
-
-    @classmethod
-    def _convert_child(cls, child):
+        import hashlib
```
Comment (vEpiphyte) [inline synapse/lib/parser.py:147]: why is this a delayed import?

File: synapse/lib/ast.py [hunk 4]
```diff
@@ -36,13 +36,30 @@ class AstNode:
...
-        self.lines = (-1, -1)
-        self.textpos = (-1, -1)
         [self.addKid(k) for k in kids]
 
+    def getAstText(self):
+        return self.astinfo.text[self.astinfo.soff:self.astinfo.eoff]
+
+    def getPosInfo(self):
+        return {
+            'lines': (self.astinfo.sline, self.astinfo.eline),
```
Comment (vEpiphyte) [inline synapse/lib/ast.py:54]: badsyntaxerror in the parser.py file includes the texthash. do we no longer have that value here?

File: synapse/lib/stormtypes.py [hunk 5]
```diff
@@ -3053,6 +3053,7 @@ async def coro():
                 raise
 
             except Exception as e:
+                import traceback; traceback.print_exc()
```
Comment (vEpiphyte) [inline synapse/lib/stormtypes.py:3056]: why is there a print here?

File: synapse/lib/view.py [hunk 6]
```diff
@@ -435,14 +435,18 @@ async def storm(self, text, opts=None):
         if editformat not in ('nodeedits', 'splices', 'count', 'none'):
             raise s_exc.BadConfValu(mesg='editformat')
 
+        import hashlib
```
Comment (vEpiphyte) [inline synapse/lib/view.py:438]: why is this a delayed import?

File: synapse/tests/test_cmds_cron.py [hunk 7]
```diff
@@ -288,20 +286,17 @@ async def getNextFoo():
                     await cmdr.runCmdLine('at --not-a-real-flag')
                     self.true(outp.expect('the following arguments'))
 
-                    await cmdr.runCmdLine('at {#foo} {#bar}')
```
Comment (vEpiphyte) [inline synapse/tests/test_cmds_cron.py:291]: why is this test removed?

File: synapse/tests/test_cmds_cron.py [hunk 8]
```diff
@@ -120,8 +120,6 @@ async def getNextFoo():
                     self.true(outp.expect('Modified cron job'))
                     await cmdr.runCmdLine(f"cron edit xxx {{[graph:node='*' :type=m2]}}")
                     self.true(outp.expect('does not match'))
-                    await cmdr.runCmdLine(f"cron mod xxx yyy")
```
Comment (vEpiphyte) [inline synapse/tests/test_cmds_cron.py:123]: is this now a valid cron command?

File: synapse/tests/test_cmds_trigger.py [hunk 9]
```diff
@@ -92,11 +92,8 @@ async def test_triggers(self):
             await cmdr.runCmdLine('trigger add')
             self.true(outp.expect('Add triggers in a cortex.'))
 
-            await cmdr.runCmdLine('trigger add tag:add #foo #bar')
```
Comment (vEpiphyte) [inline synapse/tests/test_cmds_trigger.py:95]: is this now a valid trigger command?

File: synapse/tests/test_cmds_trigger.py [hunk 10]
```diff
@@ -113,15 +110,9 @@ async def test_triggers(self):
             await cmdr.runCmdLine('trigger add prop:set test:type10.intprop test:str {test:str}')
             self.true(outp.expect('single prop'))
 
-            await cmdr.runCmdLine('trigger add tag:add #tag test:int')
```
Comment (vEpiphyte) [inline synapse/tests/test_cmds_trigger.py:116]: why was this removed?

File: synapse/tests/test_cmds_trigger.py [hunk 11]
```diff
@@ -113,15 +110,9 @@ async def test_triggers(self):
...
-            self.true(outp.expect('Missing query'))
-
             await cmdr.runCmdLine('trigger add node:add #tag1 {test:str}')
             self.true(outp.expect('Missing form'))
 
-            await cmdr.runCmdLine(f'trigger mod {goodbuid2} test:str')
-            self.true(outp.expect('start with {'))
```
Comment (vEpiphyte) [inline synapse/tests/test_cmds_trigger.py:123]: is this now valid?

File: synapse/tests/test_lib_grammar.py [hunk 12]
```diff
@@ -1269,10 +1269,10 @@ def test_cmdrargs(self):
         q = '''add {inet:fqdn | graph 2 --filter { -#nope } } inet:f-M +1 { [ graph:node='*' :type=m1]}'''
         correct = (
             'add',
-            '{inet:fqdn | graph 2 --filter { -#nope } }',
```
Comment (vEpiphyte) [inline synapse/tests/test_lib_grammar.py:1272]: is this no longer valid?

File: synapse/exc.py (hunk 2 above)
Comment (invisig0th) [reply synapse/exc.py:227]: It only implements a ctor pattern that calls the regular one so it's no more/less of an issue than we already had :+1:

File: synapse/lib/parser.py (hunk 3 above)
Comment (invisig0th) [reply synapse/lib/parser.py:147]: whups! :+1:

File: synapse/tests/test_cmds_cron.py (hunk 8 above)
Comment (invisig0th) [reply synapse/tests/test_cmds_cron.py:123]: Yep! Because the argument can be a string or an argv query.

File: synapse/tests/test_cmds_cron.py (hunk 7 above)
Comment (invisig0th) [reply synapse/tests/test_cmds_cron.py:291]: Because we tightened up the arg definitions and it's no longer a useful test cast

File: synapse/tests/test_cmds_trigger.py (hunk 11 above)
Comment (invisig0th) [reply synapse/tests/test_cmds_trigger.py:123]: Yep!

File: synapse/tests/test_lib_grammar.py (hunk 12 above)
Comment (invisig0th) [reply synapse/tests/test_lib_grammar.py:1272]: Only in the case of argv values being pre-split. The leading/trailing curly braces are no longer correct in-band.

File: synapse/tests/test_cmds_trigger.py (hunk 9 above)
Comment (invisig0th) [reply synapse/tests/test_cmds_trigger.py:95]: yep!

**Review by vEpiphyte — APPROVED** [review]
